### PR TITLE
Fix build on Haiku

### PR DESCRIPTION
**Pull request by diversys**

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -404,7 +404,7 @@ endif()
 CHECK_CXX_SOURCE_COMPILES("thread_local int i; int main() { i = 0; }"
 	HAVE_THREAD_LOCAL)
 
-if( NOT HAVE_THREAD_LOCAL )
+if( NOT HAVE_THREAD_LOCAL AND NOT HAIKU)
 	message( SEND_ERROR "C++ compiler doesn't support thread_local storage duration specifier" )
 endif()
```
Comment (alexey-lysiuk) [inline src/CMakeLists.txt]: I don’t really get why this was changed. Thread local storage CMake test fails to build but GZDoom itself compiles and links without an issue, right? If so, the test should be fixed.

File: CMakeLists.txt [hunk 2]
```diff
@@ -251,6 +251,13 @@ else()
 	endif( APPLE )
 endif()
 
+
+	if(HAIKU)
+		# Haiku networking resides in libnetwork
+		set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -lnetwork")
+		set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -lnetwork")
```
Comment (alexey-lysiuk) [inline CMakeLists.txt]: Does this really needed? We have no shared libraries. And what’s the point of the indent?

File: src/m_alloc.cpp [hunk 3]
```diff
@@ -40,6 +40,9 @@
 #include <malloc/malloc.h>
 #elif defined(__OpenBSD__)
 #include <stdlib.h>
+#elif defined(__APPLE__)
+#include <stdlib.h>
+#include <malloc.h>
 #else
```
Comment (alexey-lysiuk) [inline src/m_alloc.cpp]: How do changes Haiku relate to macOS?

File: src/CMakeLists.txt (hunk 1 above)
Comment (diversys) [reply src/CMakeLists.txt]: Without this change I get this:
```CMake Error at src/CMakeLists.txt:408 (message):
  C++ compiler doesn't support thread_local storage duration specifier
```
GCC on Haiku (as of now) is compiled with thread local storage disabled due to Internal Compiler Exception (ICE).

According to https://www.doomworld.com/forum/topic/94291-haiku-os-support/?page=2&tab=comments#comment-1812795 TLS check can be ignored.

File: CMakeLists.txt (hunk 2 above)
Comment (diversys) [reply CMakeLists.txt]: Yes, sorry, the CMAKE_SHARED_LINKER_FLAGS can probably be omitted. I just copied it from some other patch.

File: src/m_alloc.cpp (hunk 3 above)
Comment (diversys) [reply src/m_alloc.cpp]: No idea how it got into this PR. This is not needed.

**Comment by diversys** [comment]
libADLMIDI builds just fine here with no patching:
image

File: src/CMakeLists.txt (hunk 1 above)
Comment (alexey-lysiuk) [reply src/CMakeLists.txt]: TLS is now mandatory.

Still why does the test fail when GZDoom compiles successfully with several `thread_local` variables inside its source code?

**Comment by raa-eruanna** [comment]
> libADLMIDI builds just fine here with no patching:

The important thing is: does it actually work?

If it does, then the TLS check needs to be fixed, not patched around. Obviously both instances of it being used compile fine, but I have no confidence in it actually working properly.

**Comment by raa-eruanna** [comment]
Then this: https://github.com/coelckers/gzdoom/pull/598#discussion-diff-224888587R407

needs to be changed, so that this line: https://github.com/coelckers/gzdoom/blob/master/src/CMakeLists.txt#L404-L405 compiles properly and the test passes on Haiku, rather than just simply working around it for that specific system.

**Comment by coelckers** [comment]
Regarding TLS, if Haiku doesn't do it, it cannot build GZDoom. Period. Changes are underway that really require it.

**Pull request closed without merging**